v3_async_wip/tests/test_iothub_session.py

```diff
@@ -0,0 +1,1289 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for
+# license information.
+# --------------------------------------------------------------------------
+
+import asyncio
+import pytest
+import ssl
+import time
+from dev_utils import custom_mock
+from pytest_lazyfixture import lazy_fixture
+from v3_async_wip.iothub_session import IoTHubSession
+from v3_async_wip import config, models, iot_exceptions
+from v3_async_wip import connection_string as cs
+from v3_async_wip import iothub_mqtt_client as mqtt
+from v3_async_wip import sastoken as st
+from v3_async_wip import signing_mechanism as sm
+
+FAKE_DEVICE_ID = "fake_device_id"
+FAKE_MODULE_ID = "fake_module_id"
+FAKE_HOSTNAME = "fake.hostname"
+FAKE_GATEWAY_HOSTNAME = "fake.gateway.hostname"
+FAKE_URI = "fake/resource/location"
+FAKE_SHARED_ACCESS_KEY = "Zm9vYmFy"
+FAKE_SIGNATURE = "ajsc8nLKacIjGsYyB4iYDFCZaRMmmDrUuY5lncYDYPI="
+
+# TODO: aenter/aexit tests
+
+# ~~~~~ Helpers ~~~~~~
+
+
+def sastoken_generator_fn():
+    return "SharedAccessSignature sr={resource}&sig={signature}&se={expiry}".format(
+        resource=FAKE_URI, signature=FAKE_SIGNATURE, expiry=str(int(time.time()) + 3600)
+    )
+
+
+def get_expected_uri(hostname, device_id, module_id):
+    if module_id:
+        return "{hostname}/devices/{device_id}/modules/{module_id}".format(
+            hostname=hostname, device_id=device_id, module_id=module_id
+        )
+    else:
+        return "{hostname}/devices/{device_id}".format(hostname=hostname, device_id=device_id)
+
+
+# ~~~~~ Fixtures ~~~~~~
+
+# Mock out the underlying client in order to not do network operations
+@pytest.fixture(autouse=True)
+def mock_mqtt_iothub_client(mocker):
+    return mocker.patch.object(mqtt, "IoTHubMQTTClient", spec=mqtt.IoTHubMQTTClient).return_value
+
+
+# @pytest.fixture(autouse=True)
+# def mock_http_iothub_client(mocker):
+#     return mocker.patch.object(http, "IoTHubHTTPClient", spec=http.IoTHubHTTPClient).return_value
```

Comment:
nit: dead

Reply:
It's likely to be back in the next PR, but it's not like it's hard to re-implement, so I'll just remove it now.